### ISSUE TITLE
question on understanding the contrastive loss

### ISSUE THREAD

**Issue by seyeeet** [wvangansbeke/Unsupervised-Semantic-Segmentation#24]
Hi

if I understand it correctly when you compute the [contrastive loss](https://github.com/wvangansbeke/Unsupervised-Semantic-Segmentation/blob/9f01c78a59ab4af369734e13d65da55c48568de4/pretrain/utils/train_utils.py#L39) you are using some kind of a [labels that their values are the same as batch they are in](https://github.com/wvangansbeke/Unsupervised-Semantic-Segmentation/blob/aa3d0bf1dda16b7b01feae948853fbb7333ebc1d/pretrain/modules/moco/builder.py#L136) and then use CE.
q1) in normal contrastive loss the loss is forced to become zero but here we are forcing it to be the same as the labels that we gave to `sal_q `.
can you please help me understand the intuition? what if we have images from the same category in different batches, wouldnot that cause an issue? in general your feed back would be really appreciated :)

q2) just a follow up
```
logits.shape: torch.Size([107240, 264])
labels.shape: torch.Size([107240])
```
wouldnot the fact that we have `dim 1 = 264` cause any issue here? I am not getting the intuition here

**Comment by wvangansbeke** [comment]
Hi @seyeeet,

The idea here is that each mask gets a unique label. That's why we do a remapping starting from the saliency ground truth. Maybe have a look at the pseudo-code in the supplementary of the paper to understand it. The loss is actually the same as the contrastive loss, but we first have to map each pixel to its corresponding unique mask label. 

The constrastive loss can then be implemented as the CE loss over the pairwise dotproducts. So for each pixel we have a distribution over all masks in the batch and we try to maximize the similarity with the centroid of its augmented mask.

**Comment by seyeeet** [comment]
Thank you :)
